# DP: gdc: Always link with -ldl when linking with -lgphobos.

Index: b/src/gcc/d/d-spec.c
===================================================================
--- a/src/gcc/d/d-spec.c
+++ b/src/gcc/d/d-spec.c
@@ -36,6 +36,8 @@
 #define WITHLIBCXX	(1<<5)
 /* this bit is set if they did `-lc'.  */
 #define WITHLIBC	(1<<6)
+/* this bit is set if they did `-ldl'.  */
+#define DLLIB		(1<<7)
 /* This bit is set when the argument should not be passed to gcc or the backend */
 #define SKIPOPT		(1<<8)
 
@@ -75,6 +77,10 @@
 #define LIBDRUNTIME_PROFILE LIBDRUNTIME
 #endif
 
+#ifndef DL_LIBRARY
+#define DL_LIBRARY "dl"
+#endif
+
 void
 lang_specific_driver (cl_decoded_option **in_decoded_options,
 		      unsigned int *in_decoded_options_count,
@@ -123,6 +129,9 @@ lang_specific_driver (cl_decoded_option
   /* "-lstdc++" if it appears on the command line.  */
   const cl_decoded_option *saw_libcxx = 0;
 
+  /* "-ldl" if it appears on the command line.  */
+  const cl_decoded_option *saw_dl = 0;
+
   /* An array used to flag each argument that needs a bit set for
      DSOURCE, MATHLIB, WITHTHREAD, WITHLIBC or WITHLIBCXX.  */
   int *args;
@@ -139,6 +148,9 @@ lang_specific_driver (cl_decoded_option
   /* By default, we throw on the time library if we have one.  */
   int need_time = (TIME_LIBRARY[0] != '\0');
 
+  /* By default, we throw on the dl library if we have one.  */
+  int need_dl = (DL_LIBRARY[0] != '\0');
+
   /* True if we saw -static. */
   int static_link = 0;
 
@@ -248,6 +260,11 @@ lang_specific_driver (cl_decoded_option
 	      args[i] |= TIMELIB;
 	      need_time = 0;
 	    }
+	  else if (strcmp (arg, DL_LIBRARY) == 0)
+	    {
+	      args[i] |= DLLIB;
+	      need_dl = 0;
+	    }
 	  else if (strcmp (arg, "c") == 0)
 	    args[i] |= WITHLIBC;
 	  else
@@ -423,6 +440,12 @@ lang_specific_driver (cl_decoded_option
 	  saw_time = &decoded_options[i];
 	}
 
+      if (!saw_dl && (args[i] & DLLIB) && library > 0)
+	{
+	  --j;
+	  saw_dl = &decoded_options[i];
+	}
+
       if (!saw_libc && (args[i] & WITHLIBC) && library > 0)
 	{
 	  --j;
@@ -483,9 +506,6 @@ lang_specific_driver (cl_decoded_option
   /* Add `-lgphobos' if we haven't already done so.  */
   if (library > 0 && phobos)
     {
-      // Default to static linking
-      if (library == 1)
-        library = 2;
 
 #ifdef HAVE_LD_STATIC_DYNAMIC
       if (library == 3 && static_link)
@@ -510,6 +530,10 @@ lang_specific_driver (cl_decoded_option
 		       CL_DRIVER, &new_decoded_options[j]);
       added_libraries++;
       j++;
+      generate_option (OPT_l, DL_LIBRARY, 1,
+		       CL_DRIVER, &new_decoded_options[j]);
+      added_libraries++;
+      j++;
 
 #ifdef HAVE_LD_STATIC_DYNAMIC
       if (library == 3 && static_link)
@@ -580,6 +604,15 @@ lang_specific_driver (cl_decoded_option
 		       &new_decoded_options[j++]);
       added_libraries++;
     }
+
+  if (saw_dl)
+    new_decoded_options[j++] = *saw_dl;
+  else if (library > 0 && need_dl)
+    {
+      generate_option (OPT_l, DL_LIBRARY, 1, CL_DRIVER,
+		       &new_decoded_options[j++]);
+      added_libraries++;
+    }
 
   if (saw_libc)
     new_decoded_options[j++] = *saw_libc;
